archunit/src/test/java/com/tngtech/archunit/library/ArchitecturesTest.java

```diff
@@ -303,6 +312,32 @@ public void onion_architecture_gathers_all_violations() {
         assertPatternMatches(result.getFailureReport().getDetails(), expectedViolations.toPatterns());
     }
 
+    @Test
+    @DataProvider(value = {"true", "false", "null"})
+    public void onion_architecture_allows_or_rejects_empty_layers(Boolean allowEmptyLayers) {
+        OnionArchitecture architecture = onionArchitecture()
+                .domainModels(absolute("onionarchitecture.domain.model.does.not.exist"))
+                .domainServices(absolute("onionarchitecture.domain.service.not.there"))
+                .applicationServices(absolute("onionarchitecture.application.http410"));
+        if (allowEmptyLayers != null) {
+            architecture.allowEmptyLayers(allowEmptyLayers);
+        }
+
+        JavaClasses classes = new ClassFileImporter().importPackages(absolute("onionarchitecture"));
+
+        EvaluationResult result = architecture.evaluate(classes);
+        boolean expectViolation = allowEmptyLayers != Boolean.TRUE;
+        assertThat(result.hasViolation()).as("result of evaluating empty layers has violation").isEqualTo(expectViolation);
+        if (expectViolation) {
```

Comment:
I'm wondering, if it would be better to have either two tests, or specify given and the assertion in the data point. I find conditionals in assertions harder to read and harder to maintain in the long run. What do you think?
I would see as alternatives to either extract the common given into a method or pass more into the data point, e.g. pass a function pair `given{ architecture -> architecture.allowEmptyLayers(true) }` and `expected{ failureReport -> assertThat(failureReport.isEmpty()).as("failure report").isTrue(); }` as data point.

Reply:
Alternatively, we can also extract helper methods such as
* `private LayeredArchitecture aLayeredArchitectureWithEmptyLayers()`
* `private void assertFailureLayeredArchitectureWithEmptyLayers(EvaluationResult result)`

and explicitly unroll the data points into
* `@Test public void onion_architecture_rejects_empty_layers_by_default()`
* `@Test public void onion_architecture_allows_empty_layers_if_configured_to_allow()`
* `@Test public void onion_architecture_rejects_empty_layers_if_explicitly_configured_to_not_allow()`